File: src/tensorflow/ops/nn/loss/loss.cc
#include "tensorflow/cc/ops/const_op.h"
#include "loss.h"

CTCLoss::CTCLoss(tensorflow::Scope& scope, 
           tensorflow::Input inputs, 
           tensorflow::Input labels_indices, 
           tensorflow::Input labels_values, 
           tensorflow::Input sequence_length, 
           bool preprocess_collapse_repeated, 
           bool ctc_merge_repeated, 
           bool ignore_longer_outputs_than_inputs) {
      if (!scope.ok())
          return;
      auto _inputs = ::tensorflow::ops::AsNodeOut(scope, inputs);
      if (!scope.ok())
          return;
      auto _labels_indices = ::tensorflow::ops::AsNodeOut(scope, labels_indices);
      if (!scope.ok())
          return;
      auto _labels_values = ::tensorflow::ops::AsNodeOut(scope, labels_values);
      if (!scope.ok())
          return;
      auto _sequence_length = ::tensorflow::ops::AsNodeOut(scope, sequence_length);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("CTCLoss");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "CTCLoss")
                                   .Input(_inputs)
                                   .Input(_labels_indices)
                                   .Input(_labels_values)
                                   .Input(_sequence_length)
                                   .Attr("preprocess_collapse_repeated", preprocess_collapse_repeated)
                                   .Attr("ctc_merge_repeated", ctc_merge_repeated)
                                   .Attr("ignore_longer_outputs_than_inputs", ignore_longer_outputs_than_inputs)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

L2Loss::L2Loss(tensorflow::Scope& scope, 
           tensorflow::Input t, 
           tensorflow::DataType T) {
      if (!scope.ok())
          return;
      auto _t = ::tensorflow::ops::AsNodeOut(scope, t);
      ::tensorflow::Node *ret;
      const auto unique_name = scope.GetUniqueNameForOp("L2Loss");
      auto builder = ::tensorflow::NodeBuilder(unique_name, "L2Loss")
                                   .Input(_t)
                                   .Attr("T", T)
      ;
      scope.UpdateBuilder(&builder);
      scope.UpdateStatus(builder.Finalize(scope.graph(), &ret));
      if (!scope.ok()) return;
      scope.UpdateStatus(scope.DoShapeInference(ret));
      this->operation = ::tensorflow::Operation(ret);
      this->output = tensorflow::Output(ret, 0);
}

File: src/tensorflow/ops/nn/loss.nim
import ../../core

{.compile:"loss/loss.cc".}


type CTCLoss* {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0//src/tensorflow/ops/nn/loss/loss.h", importcpp:"CTCLoss/*'0*/".} = object
  operation*: Operation[ofloat]
  output*: ofloat

proc iicTCLoss(scope: Scope,
              inputs: ofloat,
              labels_indices: oint64,
              labels_values: oint32,
              sequence_length: oint32,
              preprocess_collapse_repeated: bool,
              ctc_merge_repeated: bool,
              ignore_longer_outputs_than_inputs: bool): CTCLoss {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0//src/tensorflow/ops/nn/loss/loss.h", importcpp:"CTCLoss(*#, #, #, #, #, #, #, #)", constructor.}

proc cTCLoss*(scope: Scope,
              inputs: ofloat,
              labels_indices: oint64,
              labels_values: oint32,
              sequence_length: oint32,
              preprocess_collapse_repeated: bool = false,
              ctc_merge_repeated: bool = true,
              ignore_longer_outputs_than_inputs: bool = false): CTCLoss =
  return iicTCLoss(scope,
                   inputs,
                   labels_indices,
                   labels_values,
                   sequence_length,
                   preprocess_collapse_repeated,
                   ctc_merge_repeated,
                   ignore_longer_outputs_than_inputs)

converter cTCLossToOut*(op: CTCLoss): ofloat {.inline.} = return op.output


type L2Loss*[oT:oall] {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0//src/tensorflow/ops/nn/loss/loss.h", importcpp:"L2Loss/*'0*/".} = object
  operation*: Operation[oT]
  output*: oT

proc iil2Loss[oT: oall](scope: Scope,
             t: oT,
             T: DType): L2Loss[oT] {.header:"/Users/martin//.nimble/pkgs/tensorflow-0.1.0//src/tensorflow/ops/nn/loss/loss.h", importcpp:"L2Loss(*#, #, #)", constructor.}

proc l2Loss*[oT: oall](scope: Scope,
             t: oT): L2Loss[oT] =
  return iil2Loss(scope,
                  t,
                  oT.oTF)

converter l2LossToOut*[oT: oall](op: L2Loss[oT]): oT {.inline.} = return op.output

